### PR TITLE
Autoregressive decoders refactor

### PR DESCRIPTION
**Pull request by varisd**
Refactor and code cleanup of the autoregressive.py, decoder.py and decoder/transformer.py

DecoderFeedables and DecoderHistories now contain attribute common for all derived decoders and an attribute `other` which is internal to a specific autoregressive implementation. When working with Autoregressive in general, the other should not be visible outside of the derived decoder.

There is also a clear distinction between DecoderFeedables and DecoderHistories. Only DecoderFeedables should be used when generating next output state. DecoderHistories should only record the decoding process history and should never be used to provide input for the next decoding step.
DecoderFeedables attributes are batch-major. DecoderHistories attributes are time-major. This is required for compatibility with current BeamSearchDecoder.

Also, the decoding loop was restructured to contain less repeating code. The `body` method is no longer abstract and contains the code common for all derived decoders, which follows:
- call `next_state(loop_state)` - method which is specific for each derived decoder and should only output pre-logit output-state, decoder-specific feedables and histories (for the .other attributes).
- then the output of the next_state is transformed into logits and next symbols input symbols are produced (this step should be independent of the derived decoder implementation)
- finally, new next feedables and histories are prepared (note that we feed the embedded input symbols instead of their vocabulary ids)

### CODE REVIEW

File: neuralmonkey/dataset.py [hunk 1]
```diff
@@ -95,6 +95,84 @@ def __init__(self,
 # pylint: enable=too-few-public-methods
 
 
+def _bucket_boundaries(max_length, min_length=8, length_bucket_step=1.1):
```
Comment (jlibovicky) [inline neuralmonkey/dataset.py]: Proč je tohle ve dvou různých pull requestech?

File: neuralmonkey/decoders/autoregressive.py [hunk 2]
```diff
@@ -77,20 +90,25 @@ class DecoderFeedables(NamedTuple(
         "DecoderFeedables",
         [("step", tf.Tensor),
          ("finished", tf.Tensor),
-         ("input_symbol", tf.Tensor),
-         ("prev_logits", tf.Tensor)])):
+         ("embedded_input", tf.Tensor),
+         ("other", Any)])):
     """The input of a single step of an autoregressive decoder.
 
+    The decoder should be able to decode in each step only using this.
```
Comment (jlibovicky) [inline neuralmonkey/decoders/autoregressive.py]: napsal bych "should be able to generate an output symbol only using information contained in this structure."

File: neuralmonkey/decoders/encoder_projection.py [hunk 3]
```diff
@@ -65,9 +65,10 @@ def func(train_mode: tf.Tensor,
                 "You must supply rnn_size for this type of encoder projection")
 
         en_concat = concat_encoder_projection(train_mode, None, encoders)
-        en_concat = dropout(en_concat, dropout_keep_prob, train_mode)
 
-        return tf.layers.dense(en_concat, rnn_size, name="encoders_projection")
+        return dropout(
```
Comment (jlibovicky) [inline neuralmonkey/decoders/encoder_projection.py]: Myslel jsme, že máme koncenci, že vracíme všechno bez dropoutu a dropoutujeme vstupy a ne výstupy. @jindrahelcl?

Comment (varisd) [reply neuralmonkey/decoders/encoder_projection.py]: V transformerech dropoutujeme v return statementu. Tenhle koncept se mi libi vic, tak jsem se ho drzel.

File: neuralmonkey/dataset.py (hunk 1 above)
Comment (varisd) [reply neuralmonkey/dataset.py]: Tenhle PR navazuje na ten druhy. Nevim, daji se nekde tyhle zavislosti explicitne znacit?

File: neuralmonkey/decoders/encoder_projection.py (hunk 3 above)
Comment (jlibovicky) [reply neuralmonkey/decoders/encoder_projection.py]: Hm, to asi děláme protože je to opsaný s Tensor2Tensor.

Comment (jlibovicky) [reply neuralmonkey/decoders/encoder_projection.py]: Ale v těch rekuretních modelech IMHO skoro všude dropoutujeme vstupy.

Comment (jindrahelcl) [reply neuralmonkey/decoders/encoder_projection.py]: úzus byl dělat dropout na začátku. Vždycky jsem o tom přemýšlel tak, že výstup má bejt plnohodnotnej výstup a jestli si ho chci vydropoutovat na vstupu další komponenty, má za to bejt zodpovědná ta další.

Comment (jindrahelcl) [reply neuralmonkey/decoders/encoder_projection.py]: a transformer je opsanej z t2t a tam dělaj ty jejich divný preprocesy a postprocesy, tak jsem to nechtěl zbytečně komplikovat tim, že bych to refaktoroval na náš úzus, abychom mohli ten kód porovnávat.

Comment (varisd) [reply neuralmonkey/decoders/encoder_projection.py]: Droupout by mela byt vlastnost vystupu, protoze to souvisi s trenovanim te dane vystupni vrstvy vrstvy.

Btw, zajimalo by me, jak by probihalo trenovani vrstvy, ktera leze jako vstup do dvou metod; jedna si ji dropoutne a druha ne.

File: neuralmonkey/decoders/autoregressive.py [hunk 4]
```diff
@@ -413,10 +426,87 @@ def loop_continue_criterion(self, *args) -> tf.Tensor:
...
+            loop_state = LoopState(*args)
+            feedables = loop_state.feedables
+            histories = loop_state.histories
+
+            with tf.variable_scope(self._variable_scope, reuse=tf.AUTO_REUSE):
+                output_state, dec_other, hist_other = self.next_state(
+                    loop_state)
+
+                logits = state_to_logits(output_state)
+                logits /= temperature
```
Comment (jlibovicky) [inline neuralmonkey/decoders/autoregressive.py]: Asi jsem slepej, odkud se sebere temperature?

**Comment by jindrahelcl** [comment]
> Když by to nebylo dohromady s tím batchováním, tak bych to mergoval. S tím batchováním je asi potřeba se domluvit s Jindrou, jak a kde to koliduje s jeho refaktorem.

Tohle mi teď přijde, že je asi dotáhnutější kus práce, takže bych to řešil dřív. Podívám se na to.

File: neuralmonkey/dataset.py (hunk 1 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Můžeš říct, že to nechceš mergovat do masteru, ale do jiný branche. pak se zobrazí jen změny relevantní pro tenhle PR. Po zamergování tý base branche pak můžeš u týhle změnit target na master  a rebasovat a changelist bude vypadat stejně.

File: neuralmonkey/decoders/autoregressive.py (hunk 4 above)
Comment (varisd) [reply neuralmonkey/decoders/autoregressive.py]: temperature se bere z get_body

**Comment by jlibovicky** [comment]
Rebasováno na masteru. Detailněji to prodju zítra.

File: tests/bpe.ini [hunk 5]
```diff
@@ -85,7 +85,7 @@ encoder=<encoder>
 class=decoders.decoder.Decoder
 name="decoder"
 encoders=[<encoder>]
-embedding_size=9
+embedding_size=10
```
Comment (jlibovicky) [inline tests/bpe.ini]: Jaký má důvod měnit ty dimenze v těch testech? Většinou jsem se je tam snažím psát tak, aby všechno mělo nějakej jinej rozměr, aby když ten test spadne a mluví o nějakých dimenzích, aby podle toho šlo identifikovat, rychle co je špatně. Tady koukám, že ve velkym nastavuješ rnn size stejně jako embeddingy.

File: neuralmonkey/decoders/autoregressive.py [hunk 6]
```diff
@@ -421,10 +433,87 @@ def loop_continue_criterion(self, *args) -> tf.Tensor:
...
+            has_just_finished = tf.equal(symbols, END_TOKEN_INDEX)
+            return tf.logical_or(finished, has_just_finished)
+
+        def state_to_logits(state: tf.Tensor) -> tf.Tensor:
+            logits = tf.matmul(state, self.decoding_w)
+            logits += self.decoding_b
+
+            if self.supress_unk:
+                unk_mask = tf.one_hot(
+                    UNK_TOKEN_INDEX, depth=len(self.vocabulary), on_value=-1e0)
```
Comment (jlibovicky) [inline neuralmonkey/decoders/autoregressive.py]: Ta `on_value` by měla být řádově míň, ne?

Comment (varisd) [reply neuralmonkey/decoders/autoregressive.py]: Jo, tohle je preklep. Diky, zes ho nasel (supress_unk nepouzivam, takze jsem to netestoval).

File: tests/bpe.ini (hunk 5 above)
Comment (varisd) [reply tests/bpe.ini]: Pridal jsem do decoder.Decoder podminku "self.embedding_size != self.output_dimension". Uz uplne nevim, z jakeho duvodu, mozna jako pripravu pro deep RNN decoder. Az ho dodelam, tak to stejne bude jinak

**Comment by varisd** [comment]
> Rebasováno na masteru. Detailněji to prodju zítra.

Kdyztak ten rebase prosim pushni do nejake nove vetve (treba <nazev>.rebased)

**Comment by jlibovicky** [comment]
> presun

Pozdě, už jsem tu větev přepsal. Jestli máš někde ještě lokálně to původní, tak jí můžeš někam pushnout. Ale stejně na té původní větvi sedí ještě ten další pull request, ne?

**Comment by varisd** [comment]
Jeste jsem faktorizoval obsah get_initial_loop_state do podmetod.

**Comment by varisd** [comment]
Co tady chybi? Jen opravit konfiguraky, aby prosly testy?

**Comment by jlibovicky** [comment]
Pokud k tomu @jindrahelcl nic nemá, tak jo.

**Comment by jindrahelcl** [comment]
já mam dovolenou. chtěl jsem se na to ještě podívat, protože je to docela velkej pull request.
Jestli na to nějak spěcháte, tak to zamergujte, po zběžným kouknutí večer v hotelu jsem si nevšimnul ničeho závažnýho.

**Comment by jlibovicky** [comment]
Ještě rybíz.
